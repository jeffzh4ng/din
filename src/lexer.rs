use serde::{Deserialize, Serialize};

// non-tokens:
// - comments
// - preprocessor directives
// - macros
// - whitespace: spaces, tabs, newlines

#[derive(PartialEq, Debug, Serialize, Deserialize)]
pub enum Category {
    // introductions (values)
    LiteralInt, // RE: [0-9]+
    Identifier, // RE: [a−zA−Z][a−zA−Z0−9]*

    // keywords (subset of identifiers)
    KeywordTypeInt, // RE: int
    KeywordMain,    // RE: main
    KeywordVoid,    // RE: void
    KeywordReturn,  // RE: return

    // eliminations (operations)
    Plus,  // RE: \+
    Minus, // RE: \-
    Star,  // RE: \*
    Slash, // RE: \/

    // punctuation
    LeftParen,  // RE: \(
    RightParen, // RE: \)
    LeftBrace,  // RE: \{
    RightBrace, // RE: \}
    SemiColon,  // RE: \;
}

#[derive(PartialEq, Debug, Serialize, Deserialize)]
pub struct Token {
    pub lexeme: Option<String>,
    pub category: Category,
}

// TODO: keep track of file and (col, row) for error reporting
// struct Position {}

#[derive(Default)]
pub struct Lexer {}

impl Lexer {
    pub fn scan(input: Vec<char>) -> Vec<Token> {
        let cs = Lexer::skip_whitespace(input);

        // literals and identifiers have arbitrary length
        // operations and punctuations are single ASCII characters
        match cs.as_slice() {
            [] => vec![],
            [f, r @ ..] => match f {
                '0'..='9' => Lexer::scan_int(cs),
                'a'..='z' | 'A'..='Z' => Lexer::scan_id(cs),
                '+' => {
                    let t = Token {
                        lexeme: Some(String::from("+")),
                        category: Category::Plus,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '-' => {
                    let t = Token {
                        lexeme: Some(String::from("-")),
                        category: Category::Minus,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '*' => {
                    let t = Token {
                        lexeme: Some(String::from("*")),
                        category: Category::Star,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '/' => {
                    let t = Token {
                        lexeme: Some(String::from("/")),
                        category: Category::Slash,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '(' => {
                    let t = Token {
                        lexeme: Some(String::from("(")),
                        category: Category::LeftParen,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                ')' => {
                    let t = Token {
                        lexeme: Some(String::from(")")),
                        category: Category::RightParen,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '{' => {
                    let t = Token {
                        lexeme: Some(String::from("{")),
                        category: Category::LeftBrace,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                '}' => {
                    let t = Token {
                        lexeme: Some(String::from("}")),
                        category: Category::RightBrace,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                ';' => {
                    let t = Token {
                        lexeme: Some(String::from(";")),
                        category: Category::SemiColon,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                _ => {
                    let t = Token {
                        lexeme: Some(String::from("PANIC?")),
                        category: Category::Plus,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
            },
        }
    }

    fn scan_int(input: Vec<char>) -> Vec<Token> {
        // scan_int calls skip_whitespace too to remain idempotent
        let cs: Vec<char> = Lexer::skip_whitespace(input);

        match cs.as_slice() {
            [] => vec![],
            [f, _r @ ..] => match f {
                '0'..='9' => {
                    #[rustfmt::skip]
                    let f = cs
                        .iter()
                        .take_while(|&&c| c.is_numeric())
                        .collect::<String>();

                    #[rustfmt::skip]
                    let r = cs
                        .into_iter()
                        .skip_while(|&c| c.is_numeric())
                        .collect::<Vec<_>>();

                    let t = Token {
                        lexeme: Some(f),
                        category: Category::LiteralInt,
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                _ => {
                    // panic
                    todo!()
                }
            },
        }
    }

    // TODO: support identifiers with alpha*numeric* characters after first alphabetic
    fn scan_id(input: Vec<char>) -> Vec<Token> {
        // scan_id calls skip_whitespace too to remain idempotent
        let cs: Vec<char> = Lexer::skip_whitespace(input);

        match cs.as_slice() {
            [] => vec![],
            [f, _r @ ..] => match f {
                'a'..='z' => {
                    #[rustfmt::skip]
                    let f = cs
                        .iter()
                        .take_while(|&&c| c.is_alphabetic())
                        .collect::<String>();

                    #[rustfmt::skip]
                    let r = cs
                        .into_iter()
                        .skip_while(|&c| c.is_alphabetic())
                        .collect::<Vec<_>>();

                    let keyword = match f.as_str() {
                        "int" => Some(Token {
                            lexeme: Some(String::from("int")),
                            category: Category::KeywordTypeInt,
                        }),
                        "main" => Some(Token {
                            lexeme: Some(String::from("main")),
                            category: Category::KeywordMain,
                        }),
                        "return" => Some(Token {
                            lexeme: Some(String::from("return")),
                            category: Category::KeywordReturn,
                        }),
                        _ => None,
                    };

                    let t = match keyword {
                        Some(k) => k,
                        None => Token {
                            lexeme: Some(f),
                            category: Category::Identifier,
                        },
                    };

                    std::iter::once(t).chain(Lexer::scan(r.to_vec())).collect()
                }
                _ => {
                    // panic
                    todo!()
                }
            },
        }
    }

    fn skip_whitespace(input: Vec<char>) -> Vec<char> {
        match input.as_slice() {
            [] => vec![],
            [f, r @ ..] => {
                if f.is_whitespace() {
                    Lexer::skip_whitespace(r.to_vec())
                } else {
                    input
                }
            }
        }
    }
}

#[cfg(test)]
fn vecs_match<T: PartialEq>(a: &Vec<T>, b: &Vec<T>) -> bool {
    #[rustfmt::skip]
    let matching = a
        .iter()
        .zip(b.iter())
        .filter(|&(a, b)| a == b)
        .count();

    matching == a.len() && matching == b.len()
}

#[cfg(test)]
mod test_valid {
    use insta;
    use std::fs;

    use super::*;

    #[test]
    fn hello() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/hello.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect();

        let output = Lexer::scan(input);
        insta::assert_yaml_snapshot!(output);
    }
}

#[cfg(test)]
mod test_invalid {}

#[cfg(test)]
mod test_skip_whitespace {
    use super::*;

    #[test]
    fn skip_space() {
        let input = "    7".chars().collect();
        let output: Vec<char> = Lexer::skip_whitespace(input);
        let expected_output = "7".chars().collect();

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn skip_newline() {
        let input = r#"




        7"#
        .chars()
        .collect();
        let output = Lexer::skip_whitespace(input);
        let expected_output = "7".chars().collect();

        assert!(vecs_match(&output, &expected_output))
    }
}

#[cfg(test)]
mod test_arithmetic {
    use super::*;

    #[test]
    fn simple() {
        let input = "9 + 8".chars().collect();
        let output = Lexer::scan(input);
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: Some(String::from("9")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("+")), category: Category::Plus },
            Token { lexeme: Some(String::from("8")), category: Category::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn simple_two() {
        let input = "90 + 80".chars().collect();
        let output = Lexer::scan(input);
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: Some(String::from("90")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("+")), category: Category::Plus },
            Token { lexeme: Some(String::from("80")), category: Category::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex() {
        let input = "2 + 3 * 5 - 8 / 3".chars().collect();
        let output = Lexer::scan(input);
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: Some(String::from("2")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("+")), category: Category::Plus },
            Token { lexeme: Some(String::from("3")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("*")), category: Category::Star },
            Token { lexeme: Some(String::from("5")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("-")), category: Category::Minus },
            Token { lexeme: Some(String::from("8")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("/")), category: Category::Slash },
            Token { lexeme: Some(String::from("3")), category: Category::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex_two() {
        let input = "22 + 33 * 55 - 88 / 33".chars().collect();
        let output = Lexer::scan(input);
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: Some(String::from("22")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("+")), category: Category::Plus },
            Token { lexeme: Some(String::from("33")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("*")), category: Category::Star },
            Token { lexeme: Some(String::from("55")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("-")), category: Category::Minus },
            Token { lexeme: Some(String::from("88")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("/")), category: Category::Slash },
            Token { lexeme: Some(String::from("33")), category: Category::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex_three() {
        let input = r#"
        23 +
        18 -
        45 * 2
        / 18
        "#
        .chars()
        .collect();
        let output = Lexer::scan(input);
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: Some(String::from("23")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("+")), category: Category::Plus },
            Token { lexeme: Some(String::from("18")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("-")), category: Category::Minus },
            Token { lexeme: Some(String::from("45")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("*")), category: Category::Star },
            Token { lexeme: Some(String::from("2")), category: Category::LiteralInt },
            Token { lexeme: Some(String::from("/")), category: Category::Slash },
            Token { lexeme: Some(String::from("18")), category: Category::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }
}
