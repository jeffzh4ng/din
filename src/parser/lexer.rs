use serde::{Deserialize, Serialize};

// non-tokens:
// - comments
// - preprocessor directives
// - macros
// - whitespace: spaces, tabs, newlines

// see: ARCHITECTURE.md for a more specified lexical grammar

// note: variations are explicitly typed. Collapsing categories like keywords
//       into one variant while outsourcing variation to lexeme field on Token
//       will produce more work for syntactic analysis, since lexeme : String
#[derive(PartialEq, Debug, Clone, Serialize, Deserialize)]
pub enum TokenType {
    // introductions (values)
    LiteralInt, // RE: [0-9]+
    Identifier, // RE: [a−zA−Z][a−zA−Z0−9]*

    // keywords (subset of identifiers)
    KeywordTypeInt,
    KeywordMain,
    KeywordVoid,

    // statements
    StatementReturn,

    // eliminations (operations)
    Plus,
    Minus,
    Star,
    Slash,

    // punctuation
    PuncLeftParen,
    PuncRightParen,
    PuncLeftBrace,
    PuncRightBrace,
    PuncSemiColon,
}

#[derive(PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct Token {
    pub lexeme: String,
    pub typ: TokenType,
}

// TODO: keep track of file and (col, row) for error reporting
// struct Position {}

// TODO: just filter out whitespace instead of having a helper function
pub fn scan(input: &[char]) -> Vec<Token> {
    let cs = skip_whitespace(input);

    // literals and identifiers have arbitrary length
    // operations and punctuations are single ASCII characters
    match cs {
        [] => vec![],
        [f, r @ ..] => match f {
            '0'..='9' => scan_int(cs),
            'a'..='z' | 'A'..='Z' => scan_id(cs),
            '+' => {
                let t = Token {
                    lexeme: String::from("+"),
                    typ: TokenType::Plus,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '-' => {
                let t = Token {
                    lexeme: String::from("-"),
                    typ: TokenType::Minus,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '*' => {
                let t = Token {
                    lexeme: String::from("*"),
                    typ: TokenType::Star,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '/' => {
                let t = Token {
                    lexeme: String::from("/"),
                    typ: TokenType::Slash,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '(' => {
                let t = Token {
                    lexeme: String::from("("),
                    typ: TokenType::PuncLeftParen,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            ')' => {
                let t = Token {
                    lexeme: String::from(")"),
                    typ: TokenType::PuncRightParen,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '{' => {
                let t = Token {
                    lexeme: String::from("{"),
                    typ: TokenType::PuncLeftBrace,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            '}' => {
                let t = Token {
                    lexeme: String::from("}"),
                    typ: TokenType::PuncRightBrace,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            ';' => {
                let t = Token {
                    lexeme: String::from(";"),
                    typ: TokenType::PuncSemiColon,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
            _ => {
                let t = Token {
                    lexeme: String::from("PANIC?"),
                    typ: TokenType::Plus,
                };

                std::iter::once(t).chain(scan(r)).collect()
            }
        },
    }
}

fn scan_int(input: &[char]) -> Vec<Token> {
    // scan_int calls skip_whitespace too to remain idempotent
    let cs = skip_whitespace(input);

    match cs {
        [] => vec![],
        [f, r @ ..] => match f {
            '0'..='9' => {
                #[rustfmt::skip]
                let i = r
                    .iter()
                    .take_while(|&&c| c.is_numeric())
                    .count();

                #[rustfmt::skip]
                let f = cs[..=i]
                    .iter()
                    .collect::<String>();
                let new_r = &cs[i + 1..];

                let t = Token {
                    lexeme: f,
                    typ: TokenType::LiteralInt,
                };

                std::iter::once(t).chain(scan(new_r)).collect()
            }
            _ => {
                // panic
                todo!()
            }
        },
    }
}

// TODO: support identifiers with alpha*numeric* characters after first alphabetic
fn scan_id(input: &[char]) -> Vec<Token> {
    // scan_id calls skip_whitespace too to remain idempotent
    let cs = skip_whitespace(input);

    match cs {
        [] => vec![],
        [f, r @ ..] => match f {
            'a'..='z' => {
                // Find the index where the alphabetic characters end
                #[rustfmt::skip]
                let i = r
                    .iter()
                    .take_while(|&&c| c.is_alphabetic())
                    .count();

                let f = (cs[..=i].iter()).collect::<String>();
                let new_r = &cs[i + 1..];

                let keyword = match f.as_str() {
                    "int" => Some(Token {
                        lexeme: String::from("int"),
                        typ: TokenType::KeywordTypeInt,
                    }),
                    "main" => Some(Token {
                        lexeme: String::from("main"),
                        typ: TokenType::KeywordMain,
                    }),
                    "return" => Some(Token {
                        lexeme: String::from("return"),
                        typ: TokenType::StatementReturn,
                    }),
                    _ => None,
                };

                let t = match keyword {
                    Some(k) => k,
                    None => Token {
                        lexeme: f,
                        typ: TokenType::Identifier,
                    },
                };

                std::iter::once(t).chain(scan(new_r)).collect()
            }
            _ => {
                // panic
                todo!()
            }
        },
    }
}

fn skip_whitespace(input: &[char]) -> &[char] {
    match input {
        [] => input,
        [f, r @ ..] => {
            if f.is_whitespace() {
                skip_whitespace(r)
            } else {
                input
            }
        }
    }
}

#[cfg(test)]
fn vecs_match<T: PartialEq>(a: &Vec<T>, b: &Vec<T>) -> bool {
    #[rustfmt::skip]
    let matching = a
        .iter()
        .zip(b.iter())
        .filter(|&(a, b)| a == b)
        .count();

    matching == a.len() && matching == b.len()
}

#[cfg(test)]
mod test_valid {
    use insta;
    use std::fs;

    use super::*;

    #[test]
    fn hello() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/hello.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn arithmetic_add() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/arithmetic/add.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn arithmetic_add_multi() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/arithmetic/add_multi.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn arithmetic_sub() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/arithmetic/sub.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn arithmetic_mult() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/arithmetic/mult.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn arithmetic_div() {
        #[rustfmt::skip]
        let input = fs::read("tests/valid/arithmetic/div.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let output = scan(input.as_slice());
        insta::assert_yaml_snapshot!(output);
    }
}

#[cfg(test)]
mod test_invalid {}

#[cfg(test)]
mod test_skip_whitespace {
    use super::*;

    #[test]
    fn skip_space() {
        let input = "    7".chars().collect::<Vec<_>>();
        let output = skip_whitespace(input.as_slice());
        let expected_output = "7".chars().collect::<Vec<_>>();

        assert!(vecs_match(&output.to_vec(), &expected_output))
    }

    #[test]
    fn skip_newline() {
        let input = r#"




        7"#
        .chars()
        .collect::<Vec<_>>();
        let output = skip_whitespace(input.as_slice());
        let expected_output = "7".chars().collect::<Vec<_>>();

        assert!(vecs_match(&output.to_vec(), &expected_output))
    }
}

#[cfg(test)]
mod test_arithmetic {
    use super::*;

    #[test]
    fn simple() {
        let input = "9 + 8".chars().collect::<Vec<_>>();
        let output = scan(input.as_slice());
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: String::from("9"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("+"), typ: TokenType::Plus },
            Token { lexeme: String::from("8"), typ: TokenType::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn simple_two() {
        let input = "90 + 80".chars().collect::<Vec<_>>();
        let output = scan(input.as_slice());
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: String::from("90"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("+"), typ: TokenType::Plus },
            Token { lexeme: String::from("80"), typ: TokenType::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex() {
        let input = "2 + 3 * 5 - 8 / 3".chars().collect::<Vec<_>>();
        let output = scan(input.as_slice());
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: String::from("2"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("+"), typ: TokenType::Plus },
            Token { lexeme: String::from("3"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("*"), typ: TokenType::Star },
            Token { lexeme: String::from("5"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("-"), typ: TokenType::Minus },
            Token { lexeme: String::from("8"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("/"), typ: TokenType::Slash },
            Token { lexeme: String::from("3"), typ: TokenType::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex_two() {
        let input = "22 + 33 * 55 - 88 / 33".chars().collect::<Vec<_>>();
        let output = scan(input.as_slice());
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: String::from("22"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("+"), typ: TokenType::Plus },
            Token { lexeme: String::from("33"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("*"), typ: TokenType::Star },
            Token { lexeme: String::from("55"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("-"), typ: TokenType::Minus },
            Token { lexeme: String::from("88"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("/"), typ: TokenType::Slash },
            Token { lexeme: String::from("33"), typ: TokenType::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }

    #[test]
    fn complex_three() {
        let input = r#"
        23 +
        18 -
        45 * 2
        / 18
        "#
        .chars()
        .collect::<Vec<_>>();
        let output = scan(input.as_slice());
        #[rustfmt::skip]
        let expected_output = vec![
            Token { lexeme: String::from("23"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("+"), typ: TokenType::Plus },
            Token { lexeme: String::from("18"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("-"), typ: TokenType::Minus },
            Token { lexeme: String::from("45"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("*"), typ: TokenType::Star },
            Token { lexeme: String::from("2"), typ: TokenType::LiteralInt },
            Token { lexeme: String::from("/"), typ: TokenType::Slash },
            Token { lexeme: String::from("18"), typ: TokenType::LiteralInt },
        ];

        assert!(vecs_match(&output, &expected_output))
    }
}
