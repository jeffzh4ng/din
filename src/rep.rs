#[derive(Debug, PartialEq)]
pub enum Expr {
    // introductions (values)
    Num(i128),
    Bool(bool),
    Let {
        identifier: String,
        binding: Box<Expr>,
        body: Box<Expr>,
    },
    Lambda {
        f_param: String,
        body: Box<Expr>,
    },

    // eliminations (operations)
    Binary {
        op: Op,
        l: Box<Expr>,
        r: Box<Expr>,
    },
    If {
        cond: Box<Expr>,
        then: Box<Expr>,
        els: Box<Expr>,
    },
    Var(String),
    LambdaApp {
        a_param: Box<Expr>,
        lambda: Box<Expr>, // choice: identifier or expr
    },
}

#[derive(Debug, PartialEq)]
pub enum Val {
    Num(i128),
    Bool(bool),
    Lam { param: String, body: Expr },
}

#[derive(Debug, PartialEq)]
pub enum Op {
    Add,
    Subtract,
    Multiply,
    Divide,
}
