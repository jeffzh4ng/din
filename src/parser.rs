use std::io;

use crate::lexer::{Token, TokenType};
use crate::rep::{Expr, MainFunction, Op, Program, Statement};

pub fn parse_program(tokens: Vec<Token>) -> Result<Program, io::Error> {
    let main_function = parse_function(tokens)?;
    Ok(Program { main_function })
}

fn parse_function(tokens: Vec<Token>) -> Result<MainFunction, io::Error> {
    let (_, r) = mtch(&tokens, TokenType::KeywordTypeInt)?;
    let (_, r) = mtch(r, TokenType::KeywordMain)?;
    let (_, r) = mtch(r, TokenType::PuncLeftParen)?;
    let (_, r) = mtch(r, TokenType::PuncRightParen)?;
    let (_, r) = mtch(r, TokenType::PuncLeftBrace)?;
    let (statement, r) = parse_statement(r)?;
    let (_, r) = mtch(r, TokenType::PuncRightBrace)?;

    if !tokens.is_empty() {
        // panic?
    }

    Ok(MainFunction { statement })
}

fn parse_statement(tokens: &[Token]) -> Result<(Statement, &[Token]), io::Error> {
    let (_, r) = mtch(tokens, TokenType::StatementReturn)?;
    let (expr, r) = parse_expr(r)?;
    let (_, r) = mtch(r, TokenType::PuncSemiColon)?;
    Ok((Statement::Return(expr), r))
}

fn parse_expr(tokens: &[Token]) -> Result<(Expr, &[Token]), io::Error> {
    // parse_expr(tokens); // infinite recursion if parsing left-associative grammar;
    match tokens {
        // expensive: cloning b/c of recursive call below
        [] => todo!(),
        [f, r @ ..] => match f.typ {
            TokenType::LiteralInt => {
                // TODO: assuming only plus for now.
                let mut parent = Expr::Binary {
                    op: Op::Add,
                    l: Box::new(Expr::Num(f.lexeme.parse().unwrap())), // TODO: unwrapping
                    r: Box::new(Expr::Num(-1)),                        // TODO??
                };
                let mut r_expr = r;

                while let Ok((f, r)) = mtch(r_expr, TokenType::Plus) {
                    let (f, r) = mtch(r, TokenType::LiteralInt)?;
                    let new_parent = Box::new(Expr::Binary {
                        op: Op::Add,
                        l: Box::new(Expr::Num(f.lexeme.parse::<i128>().unwrap())), // inheriting rust's i128 for now
                        r: Box::new(Expr::Num(-1)),                                // TODO??
                    });

                    parent = match parent {
                        Expr::Binary { op, l, r } => Expr::Binary {
                            op,
                            l,
                            r: new_parent,
                        },
                        _ => todo!(),
                    };

                    r_expr = r;
                }

                Ok((parent, r_expr))
            }
            TokenType::PuncLeftParen => todo!(),
            _ => todo!(), // panic?
        },
    }
}

fn parse_binop(tokens: Vec<Token>) -> Expr {
    todo!()
}

fn check(tokens: Vec<Token>, tt: TokenType) {
    todo!()
}

fn mtch<'a>(tokens: &'a [Token], tt: TokenType) -> Result<(&'a Token, &'a [Token]), io::Error> {
    match tokens {
        [] => todo!(),
        [f, r @ ..] => {
            if f.typ == tt {
                // Use an if-guard to compare values
                Ok((f, r))
            } else {
                Err(io::Error::new(io::ErrorKind::Other, "bla"))
            }
        }
    }
}

#[cfg(test)]
mod test_valid {
    use super::*;
    use crate::lexer;
    use insta;
    use std::fs;

    #[test]
    fn test_valid() {
        #[rustfmt::skip]
        let chars = fs::read("tests/valid/hello.c")
            .expect("Should have been able to read the file")
            .iter()
            .map(|b| *b as char)
            .collect::<Vec<_>>();

        let tokens = lexer::scan(&chars);
        let tree = parse_program(tokens).unwrap();
        insta::assert_yaml_snapshot!(tree);
    }
}
