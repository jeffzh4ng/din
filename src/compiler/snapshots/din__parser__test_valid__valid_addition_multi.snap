---
source: src/parser.rs
expression: tree
---
main_function:
  statement:
    Return:
      Binary:
        op: Add
        l:
          Num: 9
        r:
          Binary:
            op: Add
            l:
              Num: 10
            r:
              Num: 11

